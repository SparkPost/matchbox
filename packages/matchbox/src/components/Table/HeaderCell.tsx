import React from 'react';
import styled from 'styled-components';
import { headerCell, horizontalAlignment } from './styles';
import { padding, width, WidthProps, compose, ResponsiveValue } from 'styled-system';
import { AlignX } from '../../helpers/types';
import { Box, BoxProps } from '../Box';

type AlignXProp = ResponsiveValue<AlignX>;

const headerSystem = compose(padding, width);

const StyledHeaderCell = styled(Box)<
  BoxProps & {
    $align?: AlignXProp;
  }
>`
  ${headerCell}
  ${headerSystem}
  ${horizontalAlignment}
`;

type TableHeaderCellProps = {
  'aria-sort'?: 'none' | 'ascending' | 'descending' | 'other';
  align?: AlignXProp;
  value?: string;
  children?: React.ReactNode;
  className?: string;
  colSpan?: number;
  rowSpan?: number;
  role?: string;
  style?: React.CSSProperties;
  scope?: string;
} & WidthProps;

const HeaderCell = React.forwardRef<HTMLTableHeaderCellElement, TableHeaderCellProps>(
  function HeaderCell(
    {
      align,
      'aria-sort': ariaSort,
      value,
      children,
      className,
      colSpan,
      role,
      rowSpan,
      scope,
      style,
      width,
    },
    userRef,
  ) {
    return (
      <StyledHeaderCell
        as="th"
        p="450"
        $align={align}
        aria-sort={ariaSort}
        width={width}
        className={className}
        colSpan={colSpan}
        ref={userRef}
        role={role}
        rowSpan={rowSpan}
        scope={scope}
        style={style}
      >
        {value || children}
      </StyledHeaderCell>
    );
  },
);

HeaderCell.displayName = 'Table.HeaderCell';
export default HeaderCell;
