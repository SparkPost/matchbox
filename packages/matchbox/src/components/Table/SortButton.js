import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { ArrowUpward, ArrowDownward } from '@sparkpost/matchbox-icons';
import { focusOutline, buttonReset } from '../../styles/helpers';

const Button = styled.button`
  ${buttonReset}
  ${focusOutline()}
  
  display: flex;
  align-items: center;

  cursor: pointer;
  padding: ${({ theme }) => theme.space[100]} ${({ theme }) => theme.space[200]};
  margin-left: -${({ theme }) => theme.space[200]};
  margin-right: -${({ theme }) => theme.space[200]};
  margin-top: -${({ theme }) => theme.space[100]};
  margin-bottom: -${({ theme }) => theme.space[100]};
  border: 1px solid transparent;
  border-radius: ${({ theme }) => theme.radii[200]};

  transition: ${({ theme }) => theme.motion.duration.fast};

  &:hover {
    background: ${({ theme }) => theme.colors.blue[200]};
    color: ${({ theme }) => theme.colors.blue[700]};
    border: 1px solid ${({ theme }) => theme.colors.blue[700]};
  }
`;

const IconWrapper = styled.span`
  display: inline-block;
  margin-left: ${({ theme }) => theme.space[200]};
  color: ${({ theme }) => theme.colors.gray[700]};
  transition: ${({ theme }) => theme.motion.duration.fast};

  ${Button}:hover & {
    color: ${({ theme }) => theme.colors.blue[700]};
  }
`;

const UnsortedWrapper = styled.span`
  display: inline-block;
  position: relative;
  height: 18px;
  width: 18px;

  svg {
    position: absolute;
  }

  svg:first-child {
    top: 0px;
    left: 0px;
  }

  svg:last-child {
    bottom: -4px;
    right: 0px;
  }
`;

function Unsorted() {
  return (
    <UnsortedWrapper>
      <ArrowUpward size={14} />
      <ArrowDownward size={14} />
    </UnsortedWrapper>
  );
}

const SortButton = React.forwardRef(function SortButton(props, userRef) {
  const { children, direction, onClick, onFocus, onBlur } = props;

  const Icon = React.useMemo(() => {
    if (direction === 'asc') {
      return ArrowUpward;
    }

    if (direction === 'desc') {
      return ArrowDownward;
    }

    return Unsorted;
  }, [direction]);

  return (
    <Button onClick={onClick} onFocus={onFocus} onBlur={onBlur} ref={userRef} type="button">
      <span>{children}</span>
      <IconWrapper>
        <Icon size={18} />
      </IconWrapper>
    </Button>
  );
});

SortButton.displayName = 'Table.SortButton';
SortButton.propTypes = {
  children: PropTypes.node,
  direction: PropTypes.oneOf(['asc', 'desc', null, undefined]),
  onBlur: PropTypes.func,
  onClick: PropTypes.func,
  onFocus: PropTypes.func,
};

export default SortButton;
