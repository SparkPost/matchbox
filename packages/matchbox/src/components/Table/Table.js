import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Cell, HeaderCell, Row } from './TableElements';
import styles from './Table.module.scss';

class Table extends Component {
  static displayName = 'Table';

  static Cell = Cell;
  static HeaderCell = HeaderCell;
  static Row = Row;

  static propTypes = {
    data: PropTypes.array,
    /**
     * React node(s)
     */
    children: PropTypes.node
  };

  render() {
    const {
      children,
      data
    } = this.props;

    const dataMarkup = data
      ? <tbody>{data.map((rowData, i) => <Row rowData={rowData} key={`Row-${i}`} />)}</tbody>
      : children;

    return (
      <table className={styles.Table}>
        {dataMarkup}
      </table>
    );
  }
}

export default Table;
