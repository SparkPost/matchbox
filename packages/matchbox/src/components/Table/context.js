import React from 'react';

export const TablePaddingContext = React.createContext({});
