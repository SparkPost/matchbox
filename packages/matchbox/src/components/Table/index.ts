export { default as Table } from './Table';
export * from './Table';
export * from './Cell';
export * from './HeaderCell';
export * from './Row';
export * from './SortButton';
export * from './TotalsRow';
