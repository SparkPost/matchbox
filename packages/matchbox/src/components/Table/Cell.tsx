import React from 'react';
import styled from 'styled-components';
import { cell, horizontalAlignment } from './styles';
import { TablePaddingContext } from './context';
import {
  padding,
  PaddingProps,
  fontSize,
  width,
  WidthProps,
  compose,
  ResponsiveValue,
} from 'styled-system';
import { pick } from '../../helpers/props';
import { AlignX } from '../../helpers/types';
import { Box, BoxProps } from '../Box';

type AlignXProp = ResponsiveValue<AlignX>;

type TableCellProps = {
  value?: React.ReactNode;
  children?: React.ReactNode;
  className?: string;
  colSpan?: number;
  role?: string;
  rowSpan?: number;
  style?: React.CSSProperties;
  align?: AlignXProp;
  width?: string | number;
} & WidthProps &
  PaddingProps;

const cellSystem = compose(padding, fontSize, width);
const StyledCell = styled(Box)<
  BoxProps & {
    $align?: AlignXProp;
  }
>`
  ${cell}
  ${cellSystem}
  ${horizontalAlignment}
`;

const Cell = React.forwardRef<HTMLTableCellElement, TableCellProps>(function Cell(
  { align, value, children, className, colSpan, role, rowSpan, style, width, ...rest },
  userRef,
) {
  const paddingContext = React.useContext(TablePaddingContext);
  const paddingProps = pick(rest, padding.propNames);
  return (
    <StyledCell
      as="td"
      {...paddingContext}
      {...paddingProps}
      $align={align}
      width={width}
      fontSize={['200', null, '300']}
      className={className}
      colSpan={colSpan}
      ref={userRef}
      role={role}
      rowSpan={rowSpan}
      style={style}
    >
      {value || children}
    </StyledCell>
  );
});

Cell.displayName = 'Table.Cell';
export default Cell;
