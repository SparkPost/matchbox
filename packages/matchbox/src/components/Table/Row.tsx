import React from 'react';
import styled from 'styled-components';
import { row, verticalAlignment } from './styles';
import { ResponsiveValue } from 'styled-system';
import { AlignY } from '../../helpers/types';
import Cell from './Cell';

type AlignYProp = ResponsiveValue<AlignY>;

const StyledRow = styled.tr<{ $alignY?: AlignYProp }>`
  ${row}
  thead & th {
    vertical-align: bottom;
  }
  td {
    ${verticalAlignment}
  }
`;

export type TableRowProps = {
  className?: string;
  onClick?: React.MouseEventHandler<HTMLTableRowElement>;
  alignY?: AlignYProp;
  /**
   * @deprecated
   */
  rowData?: React.ReactNode[];
  children?: React.ReactNode;
};

const Row = React.forwardRef<HTMLTableRowElement, TableRowProps>(function Row(
  { alignY = 'center', rowData, children, className, onClick },
  userRef,
) {
  return (
    <StyledRow $alignY={alignY} className={className} onClick={onClick} ref={userRef}>
      {rowData
        ? rowData.map((value?: React.ReactNode, i?: number) => (
            <Cell value={value} key={`Cell-${i}`} />
          ))
        : children}
    </StyledRow>
  );
});

Row.displayName = 'Table.Row';
export default Row;
