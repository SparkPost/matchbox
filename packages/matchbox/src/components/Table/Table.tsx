import React from 'react';
import styled from 'styled-components';
import { margin, MarginProps, padding, PaddingProps } from 'styled-system';
import { Box } from '../Box';
import { ScreenReaderOnly } from '../ScreenReaderOnly';
import { pick } from '../../helpers/props';
import Cell from './Cell';
import HeaderCell from './HeaderCell';
import Row from './Row';
import TotalsRow from './TotalsRow';
import SortButton from './SortButton';
import { TablePaddingContext } from './context';
import { table, wrapper, sticky } from './styles';

const StyledTable = styled('table')`
  ${table}
  ${padding}
  ${sticky}
`;

const Wrapper = styled(Box)`
  ${wrapper}
  ${margin}
`;

export type TableProps = {
  'aria-readonly'?: string;
  children?: React.ReactNode;
  /**
   * @deprecated Use children instead
   */
  data?: React.ReactNode[][];
  'data-id'?: string;
  freezeFirstColumn?: boolean;
  id?: string;
  role?: string;
  title?: string;
} & MarginProps &
  PaddingProps;

function Table(props: TableProps): JSX.Element {
  const {
    'aria-readonly': readOnly,
    children,
    data,
    'data-id': dataId,
    freezeFirstColumn,
    id,
    role,
    title,
    ...rest
  } = props;
  const [isScrolled, setIsScrolled] = React.useState(false);
  const handleScroll = React.useCallback(
    (e) => {
      setIsScrolled(e.target.scrollLeft > 10);
    },
    [freezeFirstColumn],
  );
  const dataMarkup = data ? (
    <tbody>
      {data.map((rowData, i) => (
        <Row rowData={rowData} key={`Row-${i}`} />
      ))}
    </tbody>
  ) : (
    children
  );
  const { px = '450', py = '400', ...paddingProps } = pick(rest, padding.propNames);
  const marginProps = pick(rest, margin.propNames);

  return (
    <Wrapper
      $freezeFirstColumn={freezeFirstColumn}
      onScroll={freezeFirstColumn ? handleScroll : null}
      {...marginProps}
    >
      <StyledTable
        aria-readonly={readOnly}
        data-id={dataId}
        $freezeFirstColumn={freezeFirstColumn}
        id={id}
        isScrolled={isScrolled}
        role={role}
      >
        <TablePaddingContext.Provider value={{ px, py, ...paddingProps }}>
          {title && (
            <caption>
              <ScreenReaderOnly>{title}</ScreenReaderOnly>
            </caption>
          )}
          {dataMarkup}
        </TablePaddingContext.Provider>
      </StyledTable>
    </Wrapper>
  );
}

Table.Cell = Cell;
Table.HeaderCell = HeaderCell;
Table.Row = Row;
Table.TotalsRow = TotalsRow;
Table.SortButton = SortButton;
Table.displayName = 'Table';
export default Table;
