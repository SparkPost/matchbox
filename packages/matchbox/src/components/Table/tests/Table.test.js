import React from 'react';
import { shallow } from 'enzyme';

import Table from '../Table';

describe('Table', () => {
  let wrapper;

  it('renders data', () => {
    const props = {
      data: [['A', 'B', 'C'], [1, 2, 3]]
    };

    wrapper = shallow(<Table {...props}/>);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders children', () => {
    wrapper = shallow(<Table>children test</Table>);
    expect(wrapper).toMatchSnapshot();
  });

  describe('Cell', () => {
    it('renders children', () => {
      wrapper = shallow(<Table.Cell>children test</Table.Cell>);
      expect(wrapper).toMatchSnapshot();
    });

    it('renders value', () => {
      wrapper = shallow(<Table.Cell value='foo' className='test'></Table.Cell>);
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('HeaderCell', () => {
    it('renders children', () => {
      wrapper = shallow(<Table.HeaderCell>children test</Table.HeaderCell>);
      expect(wrapper).toMatchSnapshot();
    });

    it('renders value', () => {
      wrapper = shallow(<Table.HeaderCell value='foo' className='test'></Table.HeaderCell>);
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('Row', () => {
    it('renders children', () => {
      wrapper = shallow(<Table.Row>children test</Table.Row>);
      expect(wrapper).toMatchSnapshot();
    });

    it('renders row data', () => {
      wrapper = shallow(<Table.Row rowData={[1, '2', <span>3</span>]} className='test' />);
      expect(wrapper).toMatchSnapshot();
    });
  });

});
