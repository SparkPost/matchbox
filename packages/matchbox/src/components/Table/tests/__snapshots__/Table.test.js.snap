// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table Cell renders children 1`] = `
<td
  className="Cell"
>
  children test
</td>
`;

exports[`Table Cell renders value 1`] = `
<td
  className="Cell test"
>
  foo
</td>
`;

exports[`Table HeaderCell renders children 1`] = `
<th
  className="HeaderCell"
>
  children test
</th>
`;

exports[`Table HeaderCell renders value 1`] = `
<th
  className="HeaderCell test"
>
  foo
</th>
`;

exports[`Table Row renders children 1`] = `
<tr
  className="Row"
>
  children test
</tr>
`;

exports[`Table Row renders row data 1`] = `
<tr
  className="Row test"
>
  <Table.Cell
    key="Cell-0"
    value={1}
  />
  <Table.Cell
    key="Cell-1"
    value="2"
  />
  <Table.Cell
    key="Cell-2"
    value={
      <span>
        3
      </span>
    }
  />
</tr>
`;

exports[`Table renders children 1`] = `
<table
  className="Table"
>
  children test
</table>
`;

exports[`Table renders data 1`] = `
<table
  className="Table"
>
  <tbody>
    <Table.Row
      key="Row-0"
      rowData={
        Array [
          "A",
          "B",
          "C",
        ]
      }
    />
    <Table.Row
      key="Row-1"
      rowData={
        Array [
          1,
          2,
          3,
        ]
      }
    />
  </tbody>
</table>
`;
